File: tools/app/supabase/functions/_shared/server_identity.ts
export type ServerIdentityKind = 'provisional' | 'durable'

export type ServerIdentityResolution =
  | {
    status: 'ok'
    value: string | null
    changed: boolean
  }
  | {
    status: 'conflict'
    existing: string
    candidate: string
  }

export function normalizeServerIdentity(value: unknown): string | null {
  if (typeof value !== 'string') return null
  const clean = value.trim().toLowerCase()
  return clean ? clean.slice(0, 256) : null
}

export function serverIdentityKind(
  value: unknown,
): ServerIdentityKind | null {
  const clean = normalizeServerIdentity(value)
  if (!clean) return null
  return clean.startsWith('endpoint:') ? 'provisional' : 'durable'
}

export function isDurableRhythmServerIdentity(value: unknown): boolean {
  return normalizeServerIdentity(value)?.startsWith('srv-') === true
}

export function resolveServerIdentity(
  existingValue: unknown,
  candidateValue: unknown,
): ServerIdentityResolution {
  const existing = normalizeServerIdentity(existingValue)
  const candidate = normalizeServerIdentity(candidateValue)
  if (!candidate) return { status: 'ok', value: existing, changed: false }
  if (!existing) return { status: 'ok', value: candidate, changed: true }
  if (existing === candidate) {
    return { status: 'ok', value: existing, changed: false }
  }

  const existingKind = serverIdentityKind(existing)
  const candidateKind = serverIdentityKind(candidate)
  if (existingKind === 'durable' && candidateKind === 'provisional') {
    return { status: 'ok', value: existing, changed: false }
  }
  if (existingKind === 'durable' && candidateKind === 'durable') {
    return { status: 'conflict', existing, candidate }
  }
  return { status: 'ok', value: candidate, changed: true }
}

File: app/flutter/rhythm_app/lib/services/server_identity.dart
enum ServerIdentityKind { provisional, durable }

String? normalizeServerIdentity(String? value) {
  final clean = value?.trim().toLowerCase();
  return clean == null || clean.isEmpty ? null : clean;
}

ServerIdentityKind? serverIdentityKind(String? value) {
  final clean = normalizeServerIdentity(value);
  if (clean == null) return null;
  return clean.startsWith('endpoint:')
      ? ServerIdentityKind.provisional
      : ServerIdentityKind.durable;
}

bool serverIdentitiesMatch(String? left, String? right) {
  final cleanLeft = normalizeServerIdentity(left);
  final cleanRight = normalizeServerIdentity(right);
  return cleanLeft != null && cleanRight != null && cleanLeft == cleanRight;
}

bool serverIdentitiesConflict(String? left, String? right) {
  final cleanLeft = normalizeServerIdentity(left);
  final cleanRight = normalizeServerIdentity(right);
  return cleanLeft != null &&
      cleanRight != null &&
      cleanLeft != cleanRight &&
      serverIdentityKind(cleanLeft) == ServerIdentityKind.durable &&
      serverIdentityKind(cleanRight) == ServerIdentityKind.durable;
}

String? preferredServerIdentity({
  String? existing,
  String? candidate,
}) {
  final cleanExisting = normalizeServerIdentity(existing);
  final cleanCandidate = normalizeServerIdentity(candidate);
  if (cleanCandidate == null) return cleanExisting;
  if (cleanExisting == null) return cleanCandidate;
  if (serverIdentityKind(cleanExisting) == ServerIdentityKind.durable &&
      serverIdentityKind(cleanCandidate) == ServerIdentityKind.provisional) {
    return cleanExisting;
  }
  return cleanCandidate;
}
